src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -74,7 +74,7 @@ typesOf ::
   [(Ident, Expr)] ->
   m [(Ident, (Expr, Type))]
 typesOf bindingGroupType moduleName vals = do
-  ((tys, subst), warnings) <- reflectErrors . censor (const mempty) . reifyErrors . listen $ liftUnify $ do
+  ((tys, subst), warnings) <- reflectErrors . censor (const mempty) . reifyErrors . listen $ withEmptySubstitution $ do
```

Comment:
Does `withEmptySubstitution` need to go on the outside here? I just remember this code was pretty tricky to get right originally, so I want to make sure the three effects still get interleaved in the right way.